src/transform/src/fusion/mod.rs

```diff
@@ -18,3 +18,47 @@ pub mod project;
 pub mod reduce;
 pub mod top_k;
 pub mod union;
+
+use crate::TransformArgs;
+use mz_expr::MirRelationExpr;
+
+/// Fuses multiple like operators together when possible.
+#[derive(Debug)]
+pub struct Fusion;
+
+impl crate::Transform for Fusion {
+    #[tracing::instrument(
+        target = "optimizer"
+        level = "trace",
+        skip_all,
+        fields(path.segment = "fusion")
+    )]
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        use mz_expr::visit::Visit;
+        relation.visit_mut_post(&mut Self::action)?;
+        mz_repr::explain_new::trace_plan(&*relation);
+        Ok(())
+    }
+}
+
+impl Fusion {
+    /// Apply fusion action for variants we know how to fuse.
+    ///
+    /// The return value indicates a changed expression, on which we should
+    /// re-execute the transform (e.g. due to a `Negate` elision leaving an
+    /// as-yet-unexplored expression).
+    fn action(expr: &mut MirRelationExpr) {
+        match expr {
+            MirRelationExpr::Filter { .. } => filter::Filter.action(expr),
+            MirRelationExpr::Map { .. } => map::Map.action(expr),
+            MirRelationExpr::Project { .. } => project::Project.action(expr),
+            MirRelationExpr::Negate { .. } => negate::Negate.action(expr),
+            MirRelationExpr::TopK { .. } => top_k::TopK.action(expr),
```

Comment:
What's the rationale for not including the `Reduce`, `Join` and `Union` fusion rules here?

Reply:
`Join` and `Union` both do non-fusion things (filter lifting and negate pushdown, respectively). `Reduce` was a bit confusing in that I think the only thing it does is `Distinct` elimination, rather than "fusion". Seemed like we could add more things once it was clear that we understood what they did though.

Reply:
I think we can split off the negate pushdown as a separate transform that is done as part of the broader normalization pass at some point before `Fusion`. Assuming that all negations are either pushed down or lifted up as much as possible upfront, I think including the a simplified version of `Union` fusion in the new `Fusion` transform that just flattens `Union` trees makes sense.

It could be that the same line of reasoning applies to the filter lifting for `Join` fusion, but I have to look at the transform current version of the transform first.

If you agree on some of the above I will open an issue and tackle this as a follow up.

Reply:
I've added a `Union` transform (and broke out `UnionNegate` as the previous one), but I think we should eventually want to have each of these transforms (or a comment to the effect that some variant doesn't support fusion).